src/spec-node/dockerfileUtils.ts

```diff
@@ -119,6 +119,54 @@ export function findBaseImage(dockerfile: Dockerfile, buildArgs: Record<string,
 	return undefined;
 }
 
+export function findImage(image: string, dockerfile: Dockerfile, buildArgs: Record<string, string>, previousStage: Stage | undefined) {
+	const resolvedImage = replaceVariablesInImage(dockerfile, buildArgs, image, previousStage);
+	return resolvedImage;
+}
+
+function replaceVariablesInImage(dockerfile: Dockerfile, buildArgs: Record<string, string>, str: string, previousStage: Stage | undefined) {
+	return [...str.matchAll(argumentExpression)]
+		.map(match => {
+			const variable = match.groups!.variable;
+			const isVarExp = match.groups!.isVarExp ? true : false;
+			let value = findValueInImage(dockerfile, buildArgs, variable, previousStage) || '';
+			if (isVarExp) {
+				// Handle replacing variable expressions (${var:+word}) if they exist
+				const option = match.groups!.option;
+				const word = match.groups!.word;
+				const isSet = value !== '';
+				value = getExpressionValue(option, isSet, word, value);
+			}
+
+			return {
+				begin: match.index!,
+				end: match.index! + match[0].length,
+				value,
+			};
+		}).reverse()
+		.reduce((str, { begin, end, value }) => str.substring(0, begin) + value + str.substring(end), str);
+}
+
+function findValueInImage(dockerfile: Dockerfile, buildArgs: Record<string, string>, variable: string, previousStage: Stage | undefined) {
+	if (buildArgs !== undefined && variable in buildArgs) {
+		return buildArgs[variable];
+	}
+
+	if (previousStage !== undefined) {
```

Comment:
I'm not sure the previous stage should play a role here, but I might be missing something given the complexity of Dockerfile's build arg support.

I wonder if we could reuse more of the existing code to do what outdatedCommandImpl.ts needs. E.g., by adding a variation of the existing `findBaseImage`:
```ts
export function findBaseImages(dockerfile: Dockerfile, buildArgs: Record<string, string>) {
	const resolvedBaseImages = {} as Record<string, string>;
	for (const stage of dockerfile.stages) {
		const image = replaceVariables(dockerfile, buildArgs, /* not available in FROM instruction */ {}, stage.from.image, dockerfile.preamble, dockerfile.preamble.instructions.length);
		const nextStage = dockerfile.stagesByLabel[image];
		if (!nextStage) {
			resolvedBaseImages[stage.from.image] = image;
		}
	}
	return resolvedBaseImages;
}
```

outdatedCommandImpl.ts could just call this if I understand correctly.

Reply:
Amazing suggestion, thank you so much! This works